robot/impl/local_robot.go

```diff
@@ -709,12 +723,135 @@ func (r *localRobot) updateWeakDependents(ctx context.Context) {
+// excluding remote robots and services, etc from the robot's config.Config.
+func (r *localRobot) getLocalParts(ctx context.Context) ([]*referenceframe.FrameSystemPart, error) {
+	cfg, err := r.Config(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	parts := make([]*referenceframe.FrameSystemPart, 0)
+	for _, component := range cfg.Components {
+		if component.Frame == nil { // no Frame means dont include in frame system.
+			continue
+		}
+
+		if component.Name == referenceframe.World {
+			return nil, errors.Errorf("cannot give frame system part the name %s", referenceframe.World)
+		}
+		if component.Frame.Parent == "" {
+			return nil, errors.Errorf("parent field in frame config for part %q is empty", component.Name)
+		}
+		cfgCopy := &referenceframe.LinkConfig{
+			ID:          component.Frame.ID,
+			Translation: component.Frame.Translation,
+			Orientation: component.Frame.Orientation,
+			Geometry:    component.Frame.Geometry,
+			Parent:      component.Frame.Parent,
+		}
+		if cfgCopy.ID == "" {
+			cfgCopy.ID = component.Name
+		}
+		model, err := r.extractModelFrameJSON(component.ResourceName())
+		if err != nil && !errors.Is(err, referenceframe.ErrNoModelInformation) {
+			// When we have non-nil errors here, it is because the resource is not yet available.
+			// In this case, we will exclude it from the FS.
+			// When it becomes available, it will be included.
+			continue
+		}
+		lif, err := cfgCopy.ParseConfig()
+		if err != nil {
+			return nil, err
+		}
+
+		parts = append(parts, &referenceframe.FrameSystemPart{FrameConfig: lif, ModelFrame: model})
+	}
+	return parts, nil
+}
+
+func (r *localRobot) getRemoteParts(ctx context.Context) ([]*referenceframe.FrameSystemPart, error) {
+	cfg, err := r.Config(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	remoteParts := make([]*referenceframe.FrameSystemPart, 0)
+	for _, remoteName := range r.RemoteNames() {
+		// build the frame system part that connects remote world to base world
+		remoteCfg, err := getRemoteRobotConfig(remoteName, cfg)
+		if err != nil {
+			return nil, errors.Wrapf(err, "remote %s", remoteName)
+		}
+		if remoteCfg.Frame == nil { // skip over remote if it has no frame info
+			r.logger.Debugf("remote %s has no frame config info, skipping", remoteName)
+			continue
+		}
+		lif, err := remoteCfg.Frame.ParseConfig()
+		if err != nil {
+			return nil, err
+		}
+		parentName := remoteCfg.Name + "_" + referenceframe.World
+		lif.SetName(parentName)
+		remoteParts = append(remoteParts, &referenceframe.FrameSystemPart{FrameConfig: lif})
+
+		// get the parts from the remote itself
+		remote, ok := r.RemoteByName(remoteCfg.Name)
+		if !ok {
+			return nil, errors.Errorf("cannot find remote robot %s", remoteCfg.Name)
+		}
+		remoteFsCfg, err := remote.FrameSystemConfig(ctx)
+		if err != nil {
+			return nil, errors.Wrapf(err, "remote %s", remote)
+		}
+		// TODO: it jumps out to me as wrong that we are mutating the remote subparts but this is how it has been done
```

Comment:
> we are mutating the remote subparts

Are you worried about mutating the underlying `r.config.Remotes`? I don't think that will be an issue once I do RSDK-2876.

Reply:
Cool thanks, I will remove the comment and leave as is